### ISSUE TITLE
problem ：compile mxnet  with  hdfs

### ISSUE THREAD

**Issue by chengdianxuezi** [dmlc/mxnet#1722]
the logs as  follow

src/io/./hdfs_filesys.h:10:18: fatal error: hdfs.h: No such file or directory
 #include <hdfs.h>
                  ^
compilation terminated.
In file included from src/io.cc:16:0:
src/io/hdfs_filesys.h:10:18: fatal error: hdfs.h: No such file or directory
 #include <hdfs.h>
                  ^
compilation terminated.
make[1]: *** [hdfs_filesys.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: *** [io.o] Error 1
make[1]: Leaving directory `/home/xitong/ouxiangqin/mxnet/dmlc-core'
ar crv lib/libmxnet.a 
g++ -DMSHADOW_FORCE_STREAM -Wall -O3 -I/home/xitong/ouxiangqin/mxnet/mshadow/ -I/home/xitong/ouxiangqin/mxnet/dmlc-core/include -fPIC -Iinclude -msse3 -funroll-loops -Wno-unused-parameter -Wno-unknown-pragmas -DMSHADOW_USE_CUDA=0 -I/usr/local/cuda/include -DMSHADOW_USE_CBLAS=1 -DMSHADOW_USE_MKL=0 -DMSHADOW_RABIT_PS=0 -DMSHADOW_DIST_PS=0 -DMXNET_USE_OPENCV=0 -fopenmp -I/home/xitong/ouxiangqin/OpenBLAS/include -DMXNET_USE_DIST_KVSTORE -I./ps-lite/include -I/home/xitong/ouxiangqin/mxnet/deps/include -DMXNET_USE_NVRTC=0 -shared -o lib/libmxnet.so build/src/resource.o build/src/io/io.o build/src/io/iter_csv.o build/src/io/iter_image_recordio.o build/src/io/iter_mnist.o build/src/common/tblob_op_registry.o build/src/common/mxrtc.o build/src/ndarray/unary_function.o build/src/ndarray/ndarray_function.o build/src/ndarray/ndarray.o build/src/operator/block_grad.o build/src/operator/leaky_relu.o build/src/operator/embedding.o build/src/operator/cross_device_copy.o build/src/operator/pooling.o build/src/operator/cast.o build/src/operator/identity_attach_KL_sparse_reg.o build/src/operator/deconvolution.o build/src/operator/swapaxis.o build/src/operator/activation.o build/src/operator/softmax_output.o build/src/operator/convolution.o build/src/operator/cudnn_batch_norm.o build/src/operator/native_op.o build/src/operator/elementwise_binary_op.o build/src/operator/regression_output.o build/src/operator/concat.o build/src/operator/lrn.o build/src/operator/crop.o build/src/operator/slice_channel.o build/src/operator/softmax_activation.o build/src/operator/ndarray_op.o build/src/operator/operator.o build/src/operator/batch_norm.o build/src/operator/dropout.o build/src/operator/elementwise_binary_scalar_op.o build/src/operator/elementwise_sum.o build/src/operator/fully_connected.o build/src/operator/upsampling.o build/src/operator/reshape.o build/src/optimizer/sgd.o build/src/optimizer/optimizer.o build/src/engine/naive_engine.o build/src/engine/threaded_engine_pooled.o build/src/engine/threaded_engine.o build/src/engine/engine.o build/src/engine/threaded_engine_perdevice.o build/src/symbol/graph_memory_allocator.o build/src/symbol/graph_executor.o build/src/symbol/static_graph.o build/src/symbol/symbol.o build/src/storage/storage.o build/src/c_api/c_predict_api.o build/src/c_api/c_api.o build/src/c_api/c_api_error.o build/src/kvstore/kvstore.o ps-lite/build/libps.a dmlc-core/libdmlc.a -pthread -lm -L/usr/local/cuda/lib64 -L/usr/local/cuda/lib -lopenblas -lrt /usr/bin/hadoop/software/hadoop/lib/native/libhdfs.a -L/home/xitong/software/java//jre/lib/amd64/server -ljvm -Wl,-rpath=/home/xitong/software/java//jre/lib/amd64/server -L/home/xitong/ouxiangqin/OpenBLAS/lib/ /home/xitong/ouxiangqin/mxnet/deps/lib/libprotobuf-lite.a /home/xitong/ouxiangqin/mxnet/deps/lib/libzmq.a
g++: error: dmlc-core/libdmlc.a: No such file or directory
g++: error: /usr/bin/hadoop/software/hadoop/lib/native/libhdfs.a: No such file or directory
make: *** [lib/libmxnet.so] Error 1
make: *** Waiting for unfinished jobs....

**Comment by hjk41** [comment]
You need to have libhdfs to compile with HDFS support. The easiest way is
to get a Hadoop 2.0 binary and search for libhdfs in it.

On Thu, Mar 24, 2016 at 7:09 PM, beihaizai <notifications@github.com> wrote:

> the logs as follow
>
> src/io/./hdfs_filesys.h:10:18: fatal error: hdfs.h: No such file or
> directory
> #include
> ^
> compilation terminated.
> In file included from src/io.cc:16:0:
> src/io/hdfs_filesys.h:10:18: fatal error: hdfs.h: No such file or directory
> #include
> ^
> compilation terminated.
> make[1]: *** [hdfs_filesys.o] Error 1
> make[1]: *** Waiting for unfinished jobs....
> make[1]: *** [io.o] Error 1
> make[1]: Leaving directory `/home/xitong/ouxiangqin/mxnet/dmlc-core'
> ar crv lib/libmxnet.a
> g++ -DMSHADOW_FORCE_STREAM -Wall -O3
> -I/home/xitong/ouxiangqin/mxnet/mshadow/
> -I/home/xitong/ouxiangqin/mxnet/dmlc-core/include -fPIC -Iinclude -msse3
> -funroll-loops -Wno-unused-parameter -Wno-unknown-pragmas
> -DMSHADOW_USE_CUDA=0 -I/usr/local/cuda/include -DMSHADOW_USE_CBLAS=1
> -DMSHADOW_USE_MKL=0 -DMSHADOW_RABIT_PS=0 -DMSHADOW_DIST_PS=0
> -DMXNET_USE_OPENCV=0 -fopenmp -I/home/xitong/ouxiangqin/OpenBLAS/include
> -DMXNET_USE_DIST_KVSTORE -I./ps-lite/include
> -I/home/xitong/ouxiangqin/mxnet/deps/include -DMXNET_USE_NVRTC=0 -shared -o
> lib/libmxnet.so build/src/resource.o build/src/io/io.o
> build/src/io/iter_csv.o build/src/io/iter_image_recordio.o
> build/src/io/iter_mnist.o build/src/common/tblob_op_registry.o
> build/src/common/mxrtc.o build/src/ndarray/unary_function.o
> build/src/ndarray/ndarray_function.o build/src/ndarray/ndarray.o
> build/src/operator/block_grad.o build/src/operator/leaky_relu.o
> build/src/operator/embedding.o build/src/operator/cross_device_copy.o
> build/src/operator/po oling.o build/src/operator/cast.o
> build/src/operator/identity_attach_KL_sparse_reg.o
> build/src/operator/deconvolution.o build/src/operator/swapaxis.o
> build/src/operator/activation.o build/src/operator/softmax_output.o
> build/src/operator/convolution.o build/src/operator/cudnn_batch_norm.o
> build/src/operator/native_op.o build/src/operator/elementwise_binary_op.o
> build/src/operator/regression_output.o build/src/operator/concat.o
> build/src/operator/lrn.o build/src/operator/crop.o
> build/src/operator/slice_channel.o build/src/operator/softmax_activation.o
> build/src/operator/ndarray_op.o build/src/operator/operator.o
> build/src/operator/batch_norm.o build/src/operator/dropout.o
> build/src/operator/elementwise_binary_scalar_op.o
> build/src/operator/elementwise_sum.o build/src/operator/fully_connected.o
> build/src/operator/upsampling.o build/src/operator/reshape.o
> build/src/optimizer/sgd.o build/src/optimizer/optimizer.o
> build/src/engine/naive_engine.o build/src/engine/threaded_engine_pooled.o b
> uild/src /engine/threaded_engine.o build/src/engine/engine.o
> build/src/engine/threaded_engine_perdevice.o
> build/src/symbol/graph_memory_allocator.o build/src/symbol/graph_executor.o
> build/src/symbol/static_graph.o build/src/symbol/symbol.o
> build/src/storage/storage.o build/src/c_api/c_predict_api.o
> build/src/c_api/c_api.o build/src/c_api/c_api_error.o
> build/src/kvstore/kvstore.o ps-lite/build/libps.a dmlc-core/libdmlc.a
> -pthread -lm -L/usr/local/cuda/lib64 -L/usr/local/cuda/lib -lopenblas -lrt
> /usr/bin/hadoop/software/hadoop/lib/native/libhdfs.a
> -L/home/xitong/software/java//jre/lib/amd64/server -ljvm
> -Wl,-rpath=/home/xitong/software/java//jre/lib/amd64/server
> -L/home/xitong/ouxiangqin/OpenBLAS/lib/
> /home/xitong/ouxiangqin/mxnet/deps/lib/libprotobuf-lite.a
> /home/xitong/ouxiangqin/mxnet/deps/lib/libzmq.a
> g++: error: dmlc-core/libdmlc.a: No such file or directory
> g++: error: /usr/bin/hadoop/software/hadoop/lib/native/libhdfs.a: No such
> file or directory
> make: *** [lib/libmxnet.so] Error 1
> make: *** Waiting for unfinished jobs....
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/dmlc/mxnet/issues/1722>
>



-- 
HONG Chuntao
System Research Group
Microsoft Research Asia

**Comment by forcecore** [comment]
Same problem here. The problem is, I have hadoop and config.mk has no configuration to specify where hadoop is.

g++: error: /home/users/xxx/mxnet/dmlc-core/libdmlc.a: No such file or directory
g++: error: /lib/native/libhdfs.a: No such file or directory
make: *** [Makefile:218: lib/libmxnet.so] Error 1
make: *** Waiting for unfinished jobs....
g++: error: /home/users/xxx/mxnet/dmlc-core/libdmlc.a: No such file or directory
g++: error: /lib/native/libhdfs.a: No such file or directory
make: *** [Makefile:234: bin/im2rec] Error 1

$ locate libhdfs
/usr/lib/hadoop-2.7.3/lib/native/libhdfs.so
/usr/lib/hadoop-2.7.3/lib/native/libhdfs.so.0.0.0

**Comment by hjk41** [comment]
Currently the CMake script only looks for libHDFS in
$(HADOOP_HOME)//lib/native. Try to set HADOOP_HOME and compile again.

**Comment by forcecore** [comment]
Setting HADOOP_HOME worked for me, thanks!

**Comment by InesBeltaief** [comment]
I have the same problem ! how to fix it please ?

**Comment by madjam** [comment]
@InesBeltaief did you try the suggestions in this thread?

**Comment by InesBeltaief** [comment]
@madjam  yes I did and I still have the same problem
 `fatal error: hdfs.h: No such file or directory compilation terminated.`

**Comment by ThomasDelteil** [comment]
@aaronmarkham where would be the best place to add some documentation on how to build with HDFS support? Currently it seems only documented in the cinnebts of some Makefile.

**Comment by aaronmarkham** [comment]
A new FAQ page would work… and then i'd mention it in the build from source sections on pages in the /install directory too, linking back to that FAQ.

**Comment by lanking520** [comment]
@aaronmarkham any update from here?

**Comment by thvasilo** [comment]
Not sure if it works for MXNet as well, but the closest I've found that actually works (in terms of building dmlc-core with HDFS support) is this one: https://github.com/dmlc/xgboost/issues/3882#issuecomment-437251268

I'm not sure if building Hadoop from source is strictly required, I've had success in the past simply using the binaries.
